### PR TITLE
Oculus Camera App Prototype

### PR DESCRIPTION
**Pull request by natezzz**
Added basic functionality for camera app. (Take screenshots by pressing F key)

### CODE REVIEW

**Comment by tomoasleep** [comment]
Are there left tasks which you should do before we start to implement touch interaction features?
If there are not, we review your code in order to merge :+1:

**Comment by natezzz** [comment]
@tomoasleep Nope, except that I haven't tested the code because I don't have a Leap Motion at hand. If with Leap it can display screenshot correctly, then it's okay to merge.

**Comment by tomoasleep** [comment]
OK, we check and review your implementation :+1:

@Sushisushi-sandesu/leap-team というわけでよろ :pray:

File: Assets/Scripts/ScreenShotScript.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		screenShotTexture = new Texture2D(100, 100);
+		screenShot.texture = screenShotTexture;
+	}
+	
+	// Update is called once per frame
+	void Update () {
+		
+	}
+	
+	void TakeScreenShot (string path) {
```
Comment (tomoasleep) [inline Assets/Scripts/ScreenShotScript.cs]: I want you to make this method private if it is not a callback method or is not called from other classes.

Comment (natezzz) [inline Assets/Scripts/ScreenShotScript.cs]: Cannot. they are overriden methods which are to be called by Unity engine.

File: Assets/Scripts/ScreenShotScript.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		StartCoroutine (CaptureScreen (path));
+	}
+	
+	IEnumerator CaptureScreen (string path) {
+		// Wait till the last possible moment before screen rendering to hide the UI
+		yield return null;
+//		GameObject.Find("UICanvas").GetComponent<Canvas>().enabled = false;
```
Comment (tomoasleep) [inline Assets/Scripts/ScreenShotScript.cs]: I want you not to comment out code but to delete it if it is no longer needed.
If you temporarily comment out code, please write the reason comment.

File: Assets/Scripts/ScreenShotScript.cs (hunk 1 above)
Comment (tomoasleep) [inline Assets/Scripts/ScreenShotScript.cs]: Callback methods such as `Start` and `Update` are defined in superclasses, but this method `TakeScreenshot` seems not to be defined in superclasses.

http://docs.unity3d.com/ScriptReference/MonoBehaviour.html

File: ProjectSettings/ProjectSettings.asset [hunk 3]
```diff
@@ -379,7 +379,6 @@ PlayerSettings:
   XboxOnePersistentLocalStorageSize: 0
```
Comment (tomoasleep) [inline ProjectSettings/ProjectSettings.asset:379]: I want not to add this file to this commit unless this file is related to this change.

File: Assets/Scripts/ScreenShotScript.cs (hunk 2 above)
Comment (natezzz) [inline Assets/Scripts/ScreenShotScript.cs]: Ok

**Comment by ganmacs** [comment]
:eyes:  :+1:

**Comment by tomoasleep** [comment]
I think the default code format of MonoDevelop is Mono,
but it is ok to space between method names and parentheses.

**Comment by natezzz** [comment]
@tomoasleep Yes. The default is Mono style.
However I used "visual studio" style for this code.

**Comment by natezzz** [comment]
Ah. Clicked "close" by accident.

**Comment by tomoasleep** [comment]
I see :bowtie:

**Comment by tomoasleep** [comment]
I think your code style is different from Visual Studio code style in if statement and method call.
https://msdn.microsoft.com/en-us/library/ff926074.aspx

**Comment by natezzz** [comment]
@tomoasleep Yes. The Mono formatter didn't format the code completely according to the Microsoft convention. I don't know why.
Maybe we should use mono style instead?

**Comment by tomoasleep** [comment]
Looks good to me (OK to merge if another one says OK)

**Comment by tomoasleep** [comment]
Forget to check if it moves... I check it now.

**Comment by natezzz** [comment]
thank you!

On Dec 2, 2015, at 13:02, Tomoya Chiba <notifications@github.com<mailto:notifications@github.com>> wrote:

Forget to check if it moves... I check it now.

##

**Comment by tomoasleep** [comment]
I tried to build the code, but the following error occurred.

```
Assets/Scripts/ScreenShotScript.cs(3,7): error CS0246: The type or namespace name `UnityEditor' could not be found. Are you missing a using directive or an assembly reference?
```

I think the namespace UnityEditor can be used only for Editor scripts.
http://answers.unity3d.com/questions/316805/unityeditor-namespace-not-found.html
http://answers.unity3d.com/questions/576746/build-error-the-type-or-namespace-name-unityeditor.html

**Comment by tomoasleep** [comment]
If the file `/Assets/_Scenes/MainScene.unity` is used only for camera app, it should be named `CameraScene.unity`

File: Assets/_Scenes/MainScene.unity [hunk 4]
```diff
@@ -203,6 +203,80 @@ Transform:
   m_Children: []
   m_Father: {fileID: 0}
   m_RootOrder: 2
+--- !u!1 &614197765
```
Comment (tomoasleep) [inline Assets/_Scenes/MainScene.unity:206]: If the file `/Assets/_Scenes/MainScene.unity` is used only for camera app, it should be named `CameraScene.unity`.

Comment (natezzz) [inline Assets/_Scenes/MainScene.unity:206]: Ah if 'UnityEditor' is not available, which path should I use to save the screenshot?

Comment (natezzz) [inline Assets/_Scenes/MainScene.unity:206]: I see. I'll change the code and the scene name

**Comment by natezzz** [comment]
~~@Sushisushi-sandesu/all Review plz.~~
Please hold. The audio is not working correctly. Need to fix

**Comment by natezzz** [comment]
@Sushisushi-sandesu/all Review plz.

**Comment by natezzz** [comment]
Now master is merged into camera-app.
@Sushisushi-sandesu/all ready for review.

**Comment by ganmacs** [comment]
good 😎

**Comment by natezzz** [comment]
@ganmacs It is working?

**Comment by ganmacs** [comment]
Not checking it is working in my environment. i just read code.
Is it working in your environment,isn't it?

**Comment by natezzz** [comment]
Yeah it is working fine.

**Comment by ganmacs** [comment]
awsome👌 
let's merge👌

**Comment by takuti** [comment]
@natezzz from now, i will confirm to run. plz wait a minute :hand:

**Comment by natezzz** [comment]
Ok thank you

**Comment by takuti** [comment]
Anyway, should I need to build stand alone? Is running procedure same as sample app?

**Comment by natezzz** [comment]
@takuti it should work in the preview and also as a standalone. Settings are the same

**Comment by takuti** [comment]
hmm...my preview window is:

screen shot 2015-12-03 at 8 26 42 pm

There is a white screen to show captured image, but camera doesn't work by pressing F. Do you have any idea?

**Comment by natezzz** [comment]
@takuti Sigh... This is the same problem happened last time we I check the camera app with Leap Motion. It doesn't seem to capture the image taken by Leap.

I don't have a leap at hand so I cannot fix it yet.

**Comment by natezzz** [comment]
@takuti I think it happened because the image taken by Leap Motion is not part of the scene. 
Need to implement some workaround. E.g. Display the image on a canvas whenever taking screenshot etc.

**Comment by takuti** [comment]
@natezzz I see. So, how did you check if it works?

**Comment by natezzz** [comment]
@takuti It captured whatever components I put on the screen. And it also captured the warning sign after running.
So it's just the Leap's image is being ignored.

**Comment by natezzz** [comment]
@takuti if you add some buttons into the scene, it should be capturing them.

The original UnityWebcamSample worked fine because it was capturing the image on a canvas. Now there is no canvas. I think that is the problem.

**Comment by takuti** [comment]
@natezzz 

_> And it also captured the warning sign after running._

really? i wasn't able to capture that..

**Comment by takuti** [comment]
When I run that, LastScreenShot canvas emit warning as `The referenced script on this behavior is missing`

**Comment by natezzz** [comment]
@takuti Really??
In that case the problem is with the .meta files.

**Comment by takuti** [comment]
I think so. ScreenShotScript is not loaded by the canvas. I will try again from the beginning....

**Comment by natezzz** [comment]
@takuti On [this page](http://answers.unity3d.com/questions/932348/visible-or-hidden-meta-files-with-git.html), it says if the .meta files are messed up, it will report "missing scripts"

**Comment by natezzz** [comment]
@takuti Can you try go to Edit-Project settings-editor and change the meta file settings to "Visible meta files"?

**Comment by natezzz** [comment]
@takuti It would be awesome if we could meet after 輪講 to deal with this problem...

**Comment by takuti** [comment]
@natezzz Tried again from `git clone`, but it didn't work. Also, `visible meta files` shows no positive effect.

ScreenShotScript is not connected to the canvas. ---> connect ---> lots of errors like NullReferenceException

File: Assets/Scripts/ScreenShotScript.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		
+	}
+
+	void OnGUI ()
+	{
+		Event e = Event.current;
+		
+		if (e.type == EventType.KeyDown && e.keyCode == KeyCode.F) {
+			string current_time = System.DateTime.Now.ToString ().Replace ("/", "_").Replace (":", "_");
+			lastScreenShotPath = Application.temporaryCachePath + "/" + current_time + ".png";
```
Comment (takuti) [inline Assets/Scripts/ScreenShotScript.cs:32]: For some reasons, this path has a problem. After replacing `Application.temporaryCachePath` with `~/Desktop/`, the number of errors are dramatically decreased.

Comment (natezzz) [inline Assets/Scripts/ScreenShotScript.cs:32]: screen shot 2015-12-03 at 9 08 55 pm

On my machine this is how it's like.

**Comment by takuti** [comment]
!!!!

File: Assets/Scripts/ScreenShotScript.cs (hunk 5 above)
Comment (natezzz) [inline Assets/Scripts/ScreenShotScript.cs:32]: That "Application.temporaryCachePath" returns a path for temporary usage. [reference](http://qiita.com/w_yang/items/8458cd790607d14b1b36)

**Comment by natezzz** [comment]
@takuti Can you show me a screenshot of the unity console screen after pressing "F"?

**Comment by takuti** [comment]
It works (also w/ LeapMotion cam)!! 

ok

All errors are caused because,

screen shot 2015-12-03 at 9 14 35 pm

all of them are not loaded in an initial cloned repository.

**Comment by natezzz** [comment]
ああああああああああ、泣きそう @takuti

File: Assets/Scripts/ScreenShotScript.cs (hunk 5 above)
Comment (takuti) [inline Assets/Scripts/ScreenShotScript.cs:32]: Sorry, this was my misunderstanding. Path has no problem.

**Comment by natezzz** [comment]
@takuti But what caused it not to be loaded though?

But it's awesome that it worked!!

**Comment by takuti** [comment]
:tada: :tada: :tada: :tada: :sun_with_face:

**Comment by takuti** [comment]
@natezzz I don't know...are any important files describing connection between the canvas and related files in .gitignore...?

**Comment by takuti** [comment]
After solving the initial loading problem, we can merge this pull request :fearful:

**Comment by natezzz** [comment]
@takuti Yes!

However I think the .meta files in this commit are indeed broken.

**Comment by takuti** [comment]
@natezzz Now my sushi project repo works correctly, and current output of `git status` is

```
On branch camera-app
Your branch is up-to-date with 'origin/camera-app'.
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   Assets/_Scenes/CameraScene.unity
        deleted:    Assets/_Scenes/MainScene.unity.meta
        modified:   ProjectSettings/AudioManager.asset
        modified:   ProjectSettings/DynamicsManager.asset
        modified:   ProjectSettings/EditorBuildSettings.asset
        modified:   ProjectSettings/EditorSettings.asset
        modified:   ProjectSettings/GraphicsSettings.asset
        modified:   ProjectSettings/InputManager.asset
        modified:   ProjectSettings/NavMeshAreas.asset
        modified:   ProjectSettings/NetworkManager.asset
        modified:   ProjectSettings/Physics2DSettings.asset
        modified:   ProjectSettings/ProjectSettings.asset
        modified:   ProjectSettings/QualitySettings.asset
        modified:   ProjectSettings/TagManager.asset
        modified:   ProjectSettings/TimeManager.asset

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        Assets/Scripts/ScreenShotScript.cs.meta
        Assets/_Scenes/CameraScene.unity.meta
```

Hence, the reasons will be in them.

**Comment by natezzz** [comment]
Yes I think so.
Can you push your updated camera-app to this branch instead?

**Comment by natezzz** [comment]
@takuti please wait.

**Comment by natezzz** [comment]
@takuti I'll push another commit

**Comment by takuti** [comment]
@natezzz yes plz

**Comment by tomoasleep** [comment]
3 'Fix .meta problem' ...

**Comment by natezzz** [comment]
@tomoasleep Sorry. Kinda messed up a little during commit ...

**Comment by natezzz** [comment]
@tomoasleep @takuti I just "git clone" and checked. It should be working fine now. Please review.

**Comment by takuti** [comment]
I have checked on my local, and it works well by just importing Leap Motion core assets. It was able to capture a screenshot as I posted before! :+1: 

[LGTM](http://lgtm.in/i/cEfucBdVp)

**Comment by natezzz** [comment]
@takuti Nice. Now we need another reviewer and we are ready to merge. :+1: 
@Sushisushi-sandesu/all Someone please review if convenient :pray:

**Comment by tomoasleep** [comment]
@natezzz good job :tada: 
[LGTM](http://lgtm.in/i/aFqtrwJln)

**Comment by natezzz** [comment]
Thank you! @tomoasleep @takuti

**Pull request merged**